File: 0077-combinations/0077-combinations.cpp
class Solution {
public:
    vector<vector<int>> res;
    vector<int> path;
    vector<vector<int>> combine(int n, int k) {
        backtracking(1, n, k);
        return res;
    }

    void backtracking(int start, int n, int k) {
        if (path.size() == k) {
            vector<int> copy = path;
            res.push_back(copy);
            return;
        }

        for (int i = start; i <= n; i++) {
            path.push_back(i);
            backtracking(i + 1, n, k);
            path.pop_back();
        }
    }
};

File: 0077-combinations/0077-combinations.py
class Solution:
    def __init__(self):
        self.path = []
        self.res = []

    def combine(self, n: int, k: int) -> List[List[int]]:
        self.backtracking(1, k, n)
        return self.res

    def backtracking(self, start, k, n):
        if len(self.path) == k:
            self.res.append(copy.copy(self.path))
            return

        for i in range(start, n - (k - len(self.path)) + 2): # k - len(self.path)
            self.path.append(i)
            self.backtracking(i + 1, k, n)
            self.path.pop()
    